### PR TITLE
Task contact list part4 1

### PR DESCRIPTION
**Pull request by MaxGmn**

### CODE REVIEW

File: ContactList/ContactList.xcodeproj/xcshareddata/xcschemes/ContactList.xcscheme [hunk 1]
```diff
@@ -45,7 +45,10 @@
       buildConfiguration = "Debug"
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
+<<<<<<< HEAD
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList.xcodeproj/xcshareddata/xcschemes/ContactList.xcscheme]: Resolve conflicts and remove conflict markers

File: ContactList/ContactList/Model.swift [hunk 2]
```diff
@@ -15,6 +15,15 @@ struct Constants {
     static let invalidColor = UIColor.red
     static let grayColorAttribute = [NSAttributedString.Key.foregroundColor: UIColor.gray]
     static let blackColorAttribute = [NSAttributedString.Key.foregroundColor: UIColor.black]
+    static let defaultDate = "01.01.1900"
+    static let dateFormat = getDateFormatter()
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/Model.swift]: `static let dateFormat = getDateFormatter()` 
->
```
lazy var dateFormatter: DateFormatter {
<code from getDateFormatter() here>
}()
```

File: ContactList/ContactList/UpdateController.swift [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+        }
+        return array
+    }
+    
+    func getDriverLicenseNumberCell() -> Presentation {
+        let driverLicenseNumber = NSLocalizedString("DRIVER_LICENSE_NUMBER_LABLE_TITLE", comment: "Driver license number")
+        return Presentation(keyboardType: .default, placeholder: driverLicenseNumber, title: driverLicenseNumber, dataType: .text(currentPersonCopy.driverLicense), cellType: .driverLicense)
+    }
+    
+    func getCurentCell(at indexPath: IndexPath) -> UITableViewCell {
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: `getCurentCell` -> `getCell `

File: ContactList/ContactList/UpdateController.swift [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+        
+        switch indexPath.row {
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: I'd recommend you not to use hardcoded index values, but rely entry in `cells`.
```
    let presentation = cells[indexPath.row]
    switch presentation.cellType {
        case .image:
        //dequeue and return image cell
        case .driverLicenseSwitch:
        //dequeue and return switch cell
        default:
        //dequeue and return text cell
    }
````

File: ContactList/ContactList/UpdateController.swift [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+            return cell
+        }
+    }
+    
+    func updatePersonInformation(index: Int, text: String = "", switchIsOn: Bool = false, image: UIImage? = nil) -> Bool?{
+        
+        var validationResult: Bool?
+        
+        if cells[index].cellType == .driverLicenseSwitch {
+            let indexPath = IndexPath(row: index + 1, section: 0)
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: Again, try not to rely on hardcoded index, but try to calculate it dynamically
`let textfieldIndex = cells.firstIndex(where: { $0 == . driverLicense})`

File: ContactList/ContactList/UpdateController.swift [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+            let cell = tableView.dequeueReusableCell(withIdentifier: "TextTableViewCell", for: indexPath) as! TextTableViewCell
+            cell.setContent(cells[indexPath.row])
+            cell.callback = {[weak self] (cell, text) in
+                return self?.updatePersonInformation(index: indexPath.row, text: text)
+            }
+            return cell
+        }
+    }
+    
+    func updatePersonInformation(index: Int, text: String = "", switchIsOn: Bool = false, image: UIImage? = nil) -> Bool?{
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: What about changing method to
```
func updatePersonInformation(_ cellType: CellType, data: Any?)
```
You may cast `data` to necessary type basing on `cellType`

File: ContactList/ContactList/UpdateController.swift [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+//        case .driverLicenseNumber(var presentation):
+//            currentPersonCopy.driverLicense = text
+//            presentation.updateDataType(with: .text(currentPersonCopy.driverLicense))
+//            cells[index] = .driverLicenseNumber(presentation)
+//        }
+//        changeSaveButtonAvailability()
+//        return nil
+//    }
+    
+    func changeImage(for index: Int) {
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: So far you have only one image cell, so no necessity to pass `index`

File: ContactList/ContactList/UpdateController.swift [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+            choosePhotoAction.addAction(changeAction)
+            choosePhotoAction.addAction(removeAction)
+            choosePhotoAction.addAction(cancel)
+            present(choosePhotoAction, animated: true, completion: nil)
+        } else {
+            runChooseImageHandler()
+        }
+    }
+    
+    func changeCurrentImage(imageState: ImageEditState, for index: Int) {
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: `for index: Int` -> `for cellType: CellType`

**Review by oleksii-lubianyi-epam — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

File: ContactList/ContactList/UpdateController.swift [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {        
+        let cell = getCell(at: indexPath)
+        return cell
+    }
+
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let cellType = cells[indexPath.row].cellType
+        switch indexPath.row {
+        case 0:
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: Get rid of hardcoded indexes

File: ContactList/ContactList/UpdateController.swift [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+    
+    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
+        dismiss(animated: true, completion: nil)
+    }
+}
+
+
+private extension UpdateController {
+    
+    func cellRegistration() {
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: `cellRegistration` -> `registerCells`

File: ContactList/ContactList/UpdateController.swift [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        tableView.register(UINib(nibName: "ImageTableViewCell", bundle: nil), forCellReuseIdentifier: "ImageTableViewCell")
+        tableView.register(UINib(nibName: "TextTableViewCell", bundle: nil), forCellReuseIdentifier: "TextTableViewCell")
+        tableView.register(UINib(nibName: "SwitchTableViewCell", bundle: nil), forCellReuseIdentifier: "SwitchTableViewCell")
+    }
+    
+    func changeSaveButtonAvailability() {
+        saveButton.isEnabled = (!currentPersonForEditing.isEqual(currentPersonCopy) || imageState != ImageEditState.noChanges) &&
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: Condition is too long to read and understand. Consider refactoring

File: ContactList/ContactList/UpdateController.swift [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    func updatePersonInformation(_ cellType: CellType, data: Any?) -> Bool?{
+        
+        guard let index = cells.firstIndex(where: {$0.cellType == cellType}) else {
+            return nil
+        }
+        
+        var validationResult: Bool?
+        
+        if cellType == .driverLicenseSwitch {
+            if data is Bool && data as! Bool {
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]:
```
let isOn = data as! Bool

if isOn {
...
} else {
...
}
```

File: ContactList/ContactList/UpdateController.swift [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+                    cells.remove(at: indexPath.row)
+                    tableView.deleteRows(at: [indexPath], with: .automatic)
+                    currentPersonCopy.driverLicense = ""
+                }
+            }
+        } else {
+            cells[index].updateDataType(cells[index].dataType, with: data)
+            
+            if let validationType = cells[index].validationType {
+                if data is String{
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: `if let newValue = data as? String`

File: ContactList/ContactList/UpdateController.swift [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+                    fieldsValidationResult[index] = Validation.isValidField(text: data as! String, kindOfField: validationType)
+                    validationResult = fieldsValidationResult[index]
+                }
+            }
+            
+            currentPersonCopy.setValue(data, forKey: cellType.rawValue)
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift:218]: I think you must check data type before updating model

File: ContactList/ContactList/UpdateController.swift [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+            
+            switch cellType {
+            case .image, .notes:
+                tableView.reloadRows(at: [IndexPath(row: index, section: 0)], with: .automatic)
+            case .birthday:
+                tableView.endEditing(true)
```
Comment (oleksii-lubianyi-epam) [inline ContactList/ContactList/UpdateController.swift]: Table should end editing only on Done button tap

**Review by oleksii-lubianyi-epam — CHANGES_REQUESTED** [review]

**Review by oleksii-lubianyi-epam — APPROVED** [review]

**Pull request merged**